Source/Lib/Codec/EbMalloc.c

```diff
@@ -0,0 +1,393 @@
+{
+    (void)InitOnce;
+    (void)Parameter;
+    (void)lpContext;
+    gMallocMutex = EbCreateMutex();
+    return TRUE;
+}
+
+static EB_HANDLE GetMallocMutex()
+{
+    InitOnceExecuteOnce(&gMallocOnce, CreateMallocMutex, NULL, NULL);
+    return gMallocMutex;
+}
+#else
+#include <pthread.h>
+static void CreateMallocMutex()
+{
+    gMallocMutex = EbCreateMutex();
+}
+
+static pthread_once_t gMallocOnce = PTHREAD_ONCE_INIT;
+
+static EB_HANDLE GetMallocMutex()
+{
+    pthread_once(&gMallocOnce, CreateMallocMutex);
+    return gMallocMutex;
+}
+#endif // _WIN32
+
+//hash function to speedup etnry search
+uint32_t hash(void* p)
+{
+#define MASK32 ((((uint64_t)1)<<32)-1)
+
+    uint64_t v = (uint64_t)p;
+    uint64_t low32 = v & MASK32;
+    return (uint32_t)((v >> 32) + low32);
+}
+
+typedef struct MemoryEntry{
+    void* ptr;
+    EB_PTRType type;
+    size_t count;
+    const char* file;
+    uint32_t line;
+} MemoryEntry;
+
+//+1 to get a better hash result
+#define MEM_ENTRY_SIZE (4 * 1024 * 1024 + 1)
+
+MemoryEntry gMemEntry[MEM_ENTRY_SIZE];
+
+#define TO_INDEX(v) ((v) % MEM_ENTRY_SIZE)
+static EB_BOOL gAddMemEntryWarning = EB_TRUE;
+static EB_BOOL gRemoveMemEntryWarning = EB_TRUE;
+
+
+/*********************************************************************************
+*
+* @brief
+*  compare and update current memory entry.
+*
+* @param[in] e
+*  current memory entry.
+*
+* @param[in] param
+*  param you set to ForEachMemEntry
+*
+*
+* @returns  return EB_TRUE if you want get early exit in ForEachMemEntry
+*
+s*
+********************************************************************************/
+
+typedef EB_BOOL (*Predicate)(MemoryEntry* e, void* param);
+
+/*********************************************************************************
+*
+* @brief
+*  Loop through mem entries.
+*
+* @param[in] bucket
+*  the hash bucket
+*
+* @param[in] start
+*  loop start position
+*
+* @param[in] pred
+*  return EB_TRUE if you want early exit
+*
+* @param[out] param
+*  param send to pred.
+*
+* @returns  return EB_TRUE if we got early exit.
+*
+*
+********************************************************************************/
+static EB_BOOL ForEachHashEntry(MemoryEntry* bucket, uint32_t start, Predicate pred, void* param)
+{
+
+    uint32_t s = TO_INDEX(start);
+    uint32_t i = s;
+
+    do {
+        MemoryEntry* e = bucket + i;
+        if (pred(e, param)) {
+            return EB_TRUE;
+        }
+        i++;
+        i = TO_INDEX(i);
+    } while (i != s);
+     return EB_FALSE;
+}
+
+static EB_BOOL ForEachMemEntry(uint32_t start, Predicate pred, void* param)
+{
+    EB_BOOL ret;
+    EB_HANDLE m = GetMallocMutex();
+    EbBlockOnMutex(m);
+    ret = ForEachHashEntry(gMemEntry, start, pred, param);
+    EbReleaseMutex(m);
+    return ret;
+}
+
+static const char* MemTypeName(EB_PTRType type)
+{
+    static const char *name[EB_PTR_TYPE_TOTAL] = {"malloced memory", "calloced memory", "aligned memory", "mutex", "semaphore", "thread"};
```

Comment:
Mutex, semaphore and thread are not memory type, although they're managed with memory together here...

Reply:
done